habitat-baselines/habitat_baselines/rl/hrl/hl/planner_policy.py

```diff
@@ -0,0 +1,316 @@
+                    # Unfortunately holding and not_holding are negations. The
+                    # PDDL system does not currently support negations, so we
+                    # have to manually handle this case.
+                    if p.name == "holding":
+                        pred_set = [
+                            other_p
+                            for other_p in pred_set
+                            if not (
+                                other_p.name == "not_holding"
+                                and other_p._arg_values[0] == p._arg_values[1]
+                            )
+                        ]
+                    if p.name == "not_holding":
+                        pred_set = [
+                            other_p
+                            for other_p in pred_set
+                            if not (
+                                other_p.name == "holding"
+                                and p._arg_values[0] == other_p._arg_values[1]
+                            )
+                        ]
+                    if p not in pred_set:
+                        pred_set.append(p)
+
+                pred_hash = _get_pred_hash(pred_set)
+
+                if pred_hash not in visited:
+                    visited.add(pred_hash)
+                    add_node = PlanNode(
+                        pred_set, cur_node, cur_node.depth + 1, action
+                    )
+                    if pddl_goal.is_true_from_predicates(pred_set):
+                        # Found a goal, we can stop searching.
+                        sol_nodes.append(add_node)
+                    else:
+                        stack.append(add_node)
+        return sol_nodes
+
+    def _extract_paths(self, sol_nodes):
+        paths = []
+        for sol_node in sol_nodes:
+            cur_node = sol_node
+            path = []
+            while cur_node.parent is not None:
+                path.append(cur_node)
+                cur_node = cur_node.parent
+            paths.append(path[::-1])
+        return paths
+
+    def _get_plan(self, pred_vals, pddl_goal: LogicalExpr) -> List[PddlAction]:
+        """
+        :param pred_vals: Shape (num_prds,). NOT batched.
+        """
+        assert len(pred_vals) == len(self._predicates_list)
+        sol_nodes = self._get_solution_nodes(pred_vals, pddl_goal)
+
+        # Extract the sequence of actions that lead to the goal.
+        paths = self._extract_paths(sol_nodes)
+
+        all_ac_seqs = []
+        for path in paths:
+            all_ac_seqs.append([node.action for node in path])
+        # Sort by the length of the action sequence
+        full_plans = sorted(all_ac_seqs, key=len)
+        # Each full plan will be a permutation of the other full plans.
+        plans = full_plans[0]
+        return plans
+
+    def _replan(self, pred_vals, plan_idx: int):
+        if self._select_random_goal:
+            # We select a plan at random
+            index_plan = plan_idx
```

Comment:
how is this a random plan? Seems like the input plan index right?

Reply:
The plan index is randomly generated. I renamed this variable to be clearer.